sim-validator-assignment/src/config.rs

```diff
@@ -93,31 +99,70 @@ impl Config {
 
         Ok(shard_seats)
     }
+
+    /// Collect partials seats for `shard_idx` by picking seats from positions with `position %
+    /// num_shards == shard_idx`.
+    ///
+    /// # Motivation for assignment via modulo
+    ///
+    /// Every validator may hold at most 1 partial seat. If a validator's stake covers full seats
+    /// without remainder, it holds 0 partial seats. Hence `0 <= num_partial_seats <=
+    /// num_validators`. Assigning seats to shards with `%` allows distributing the existing number
+    /// of partial seats to shards as evenly as possible.
```

Comment:
This choice is probably ok for the simulation (I don't think it will impact the results), but for production I think we need to be more careful about this assignment. While using the modulus is even, it is also biased to smaller shard ids (for example suppose there are 4 shards and 9 partial seats, then shard 0 gets 3 partial seats while all the others get 2). This means it will be systematically easier to corrupt higher shard ids than earlier ones. To even this out we probably need to randomly shuffle the shard order as well as the partial seats. By "shuffle the shard order" I mean we can still use this scheme to assign an index to each partial seat, but then that index may or may not equal the shard index, depending on how we have decided to permute the shards (for example [0, 1, 2, 3] might be shuffled to [1, 3, 2, 0]).

Another idea could be to correlate partial and full seat assignment to try to even out stake as much as possible. For example, suppose there are 4 shards, 13 seats and 9 partial seats. With the current scheme shard 0 is getting the extra full seat and the extra partial seat. If there is a partial seat that is "close" to being a full seat it should be separated from the other partial seats and not in the same shard as the extra full seat because that would result in the most even stake distribution.

Making that correlation rigorous could be challenging and may impact the security properties of the algorithm. Perhaps it is worth thinking about more carefully.

Reply:
> This choice is probably ok for the simulation (I don't think it will impact the results)

The result only counts the number of corrupted shards, so I also think it’s fine for the simulation.

> While using the modulus is even, it is also biased to smaller shard ids 

Good point! Shuffling the shard order sounds good to me since it is a relatively cheap operation. Especially since the number of shards should remain low in the proximate future.

> Another idea could be to correlate partial and full seat assignment to try to even out stake as much as possible.

A concern could be its computational and logical overhead. I think a high number of seats is preferred as it loweres the probability of shard corrruption. A high number of seats implies low `stake_per_seat` and overall less stake assigned to partial seats. So it might also need to be considered how much safety the correlations adds vs how much costs the overhead causes, since the compuation of an assignment is required at every height.

A middle ground could be the shuffling of the shard order for _both_ full and partial seats. This should at least remove the bias towards smaller shard ids.

Reply:
Yeah, the shard order independently for shuffling both full and partial seats sounds like a good idea.